### ISSUE TITLE
Exceedingly high CPU usage on Windows

### ISSUE THREAD

**Issue by webarchymeta** [node-webrtc/node-webrtc#340]
OS: Windows 10 (1703).

When version 0.0.63 is used, the percentage of CPU usage by a single node process running wrtc remains at a level of 30% to 45% (CPU: Intel i5, 4 cores). This is not the case when 0.0.62 is used. The CPU usage remains quite low even on a NCU (next computing unit with low end Intel Atom CPU)

It looks normal on Linux systems, even for virtual machines ...

[Update]:
It seems to yield well to other processes when they need to use the CPU. But when the system is otherwise idle, its CPU usage level returns to high. This could generates potential heat problems on small devices ...

**Comment by charllie** [comment]
Same here on Windows 7

0.0.62: no high CPU
0.0.63: high CPU usage
0.0.64: high CPU usage

**Comment by markandrus** [comment]
Thanks for the report. I can reproduce. I think it's the FakeAudioDeviceModule.

**Comment by markandrus** [comment]
Actually, I think the issue is primarily the multiple PeerConnectionFactoryInterfaces. Here is my writeup. I'll followup with a PR soon. We can also make the test mentioned in the writeup part of our Travis CI tests to ensure we don't regress again.

---

# Background

node-webrtc needs to create a PeerConnectionFactoryInterface (PCFI) in order to create the underlying PeerConnectionInterfaces (PCIs) used by the library. Historically, node-webrtc has created one PCFI per PCI.

node-webrtc also has the option to provide an AudioDeviceModule (ADM) to the PCFI. If we don't provide one, the platform default is used. On Linux, this default is PulseAudio or ALSA, which causes problems if node-webrtc is used where these are unavailable.

node-webrtc v0.0.62 was the last release to use the default ADM, and, in v0.0.63, we started using FakeAudioDeviceModule (FADM). There was a [warning on the PR](https://github.com/js-platform/node-webrtc/pull/287#issuecomment-242502411) that this increased CPU usage, but I didn't really realize how much CPU usage increased until I started looking into #340 and #351.

# Analysis

It turns out, there are two changes we can introduce to reduce CPU usage:

1. We can start sharing a PCFI among PCIs.
2. Instead of using FADM, we can create a "dummy" ADM with

    ```cpp
    webrtc::AudioDeviceModule::Create(0, webrtc::AudioDeviceModule::AudioLayer::kDummyAudio);
    ```

I validated this with a [test](#test) using different versions of node-webrtc on an m5.xlarge (4 cores) running Ubuntu. I don't have a good way to measure Chrome, but node-webrtc seems to perform as quickly as Chrome with the recommended changes.

| node-webrtc | WebRTC | ADM | PCFI | Avg. CPU (%) | Max CPU (%) | Time (s) | 
|:--- |:--- |:--- |:--- |:--- |:--- |:--- |
| [v0.0.62](#v0062) | M50 | default | N | 110.891 | 149.4 | 56.877 |
| [v0.0.63–v0.0.65](#v0063v0065) | M57 | FADM | N | * | * | * |
| [develop](#develop-with-fadm-and-1-pcfi) | M57 | FADM | 1 | 95.842 | 149.9 | 6.644 |
| [develop](#develop-with-dummy-adm-and-n-pcfis) | M57 | dummy | N | * | * | * |
| [develop](#develop-with-dummy-adm-and-1-pcfi) | M57 | dummy | 1 | 81.492 | 125.24 | 6.367 |

\* I can't actually complete the tests in these versions. node-webrtc takes more and more CPU, topping out at ~84 RTCPeerConnections.

# Test

The test creates 250 pairs of RTCPeerConnections. Each pair negotiates an RTCDataChannel. Then one side of the RTCDataChannel is used to send a "ping", and the other side is used to send a "pong". The test can also be run in the browser, where it will add a button, "Test", to the page.

<details><summary>test.js</summary>
<p>

```js
'use strict';

if (typeof require === 'function') {
  Object.assign(global, require('wrtc'));
}

function createRTCPeerConnections() {
  const pc1 = new RTCPeerConnection();
  const pc2 = new RTCPeerConnection();
  [[pc1, pc2], [pc2, pc1]].forEach(([pc1, pc2], i) => {
    pc1.onicecandidate = async ({ candidate }) => {
      if (candidate) {
        await pc2.addIceCandidate(candidate);
        return;
      }
    };
  });
  return [pc1, pc2];
}

async function negotiate(offerer, answerer) {
  const offer = await offerer.createOffer();
  await Promise.all([
    offerer.setLocalDescription(offer),
    answerer.setRemoteDescription(offer)
  ]);
  const answer = await answerer.createAnswer();
  await Promise.all([
    offerer.setRemoteDescription(answer),
    answerer.setLocalDescription(answer)
  ]);
}

function receive(dc) {
  return Promise.race([
    new Promise(resolve =>
      dc.onmessage = ({ data }) => resolve(data)),
    new Promise((resolve, reject) =>
      dc.onclose = () => reject(new Error('RTCDataChannel closed'))),
    new Promise((resolve, reject) =>
      dc.onerror = ({ message }) => reject(new Error(message))),
  ]);
}

async function sendAndReceive(dc1, dc2, input) {
  if (dc1.readyState === 'connecting') {
    await new Promise(resolve => dc1.onopen = resolve);
  } else if (dc1.readyState !== 'open') {
    throw new Error(`RTCDataChannel is in readyState "${readyState}"`);
  }

  dc1.send(input);

  const output = await Promise.race([
    receive(dc2),
    new Promise((resolve, reject) =>
      dc1.onclose = () => reject(new Error('RTCDataChannel closed'))),
    new Promise((resolve, reject) =>
      dc1.onerror = ({ message }) => reject(new Error(message)))
  ]);

  if (input !== output) {
    throw new Error(`Expected "${input}", got "${output}"`);
  }
}

class RTCPeerConnectionPair {
  constructor() {
    const [pc1, pc2] = createRTCPeerConnections();

    this.pc1 = pc1;
    this.pc2 = pc2;

    this.dc1 = pc1.createDataChannel('ping-pong');

    this.dc2 = null;
    this.dc2Promise = new Promise(resolve =>
      pc2.ondatachannel = ({ channel }) => resolve(channel));
  }

  async open() {
    await negotiate(this.pc1, this.pc2);

    await this.dc1Promise;

    this.dc2 = await this.dc2Promise;

    await sendAndReceive(this.dc1, this.dc2, 'ping');
    await sendAndReceive(this.dc2, this.dc1, 'pong');
  }

  close() {
    this.dc1.close();
    this.dc2.close();
    this.pc1.close();
    this.pc2.close();
  }
}

async function test(n) {
  const pairs = [];
  try {
    for (let i = 0; i < n; i++) {
      console.log(`Creating RTCPeerConnectionPair ${i+1}`);
      const pair = new RTCPeerConnectionPair();
      await pair.open();
      pairs.push(pair);
    }
  } catch (error) {
    throw error;
  } finally {
    console.log(`Closing ${pairs.length} RTCPeerConnectionPair(s)`);
    pairs.forEach(pair => pair.close());
  }
}

function main() {
  return test(250);
}

if (typeof require === 'function') {
  if (require.main === module) {
    main().catch(error => {
      console.error(error.stack);
      process.exit(1);
    });
  } else {
    module.exports = test;
  }
} else if (typeof require === 'undefined' && typeof document === 'object') {
  const button = document.createElement('button');
  button.innerText = 'Test';
  button.onclick = async event => {
    event.preventDefault();
    button.disabled = true;
    try {
      await main();
    } catch (error) {
      throw error;
    } finally {
      button.disabled = false;
    }
  };
  document.body.appendChild(button);
}
```

</p>
</details>

## Measuring CPU Usage and Time

```bash
time (node test.js >/dev/null & PID=$! ; while ps -p $PID -o '%cpu'; do true; done) \
  | grep -v ' *%CPU' \
  | awk 'BEGIN { total = max = 0 } { total += $0; if ($0 > max) max = $0 } END { print (total/NR), max }'
```

# Results

Results were averaged over 10 runs.

## v0.0.62

* Average CPU Usage (%): 110.891
* Max CPU Usage (%): 149.4
* Time (s): 56.877

## v0.0.63–v0.0.65

There aren't many changes between v0.0.63 and v0.0.65, so I tested v0.0.65. I can't actually complete the test. node-webrtc takes more and more CPU, topping out at ~84 RTCPeerConnections.

## develop with FADM and 1 PCFI

<details><summary>Patch</summary>
<p>
  
```patch
diff --git a/src/binding.cc b/src/binding.cc
index 6109516..4000ab8 100644
--- a/src/binding.cc
+++ b/src/binding.cc
@@ -38,6 +38,8 @@ void setup(rtc::Thread* signalingThread, rtc::Thread* workerThread) {
 }
 void dispose(void* args) {
+  node_webrtc::PeerConnection::Dispose();
+
   workerThread.Stop();
   signalingThread.Stop();
diff --git a/src/peerconnection.cc b/src/peerconnection.cc
index b112f69..721ca32 100644
--- a/src/peerconnection.cc
+++ b/src/peerconnection.cc
@@ -10,7 +10,7 @@
 #include "webrtc/api/mediaconstraintsinterface.h"
 #include "webrtc/api/test/fakeconstraints.h"
 #include "webrtc/base/refcount.h"
-#include "webrtc/modules/audio_device/include/fake_audio_device.h"
+#include "webrtc/modules/audio_device/include/audio_device.h"
 #include "common.h"
 #include "create-answer-observer.h"
@@ -36,6 +36,7 @@ using v8::Value;
 using v8::Array;
 Nan::Persistent<Function> PeerConnection::constructor;
+rtc::scoped_refptr<webrtc::PeerConnectionFactoryInterface> PeerConnection::_jinglePeerConnectionFactory;
 rtc::Thread* PeerConnection::_signalingThread;
 rtc::Thread* PeerConnection::_workerThread;
@@ -59,8 +60,11 @@ PeerConnection::PeerConnection(webrtc::PeerConnectionInterface::IceServers iceSe
   constraints.AddMandatory(webrtc::MediaConstraintsInterface::kOfferToReceiveAudio, webrtc::MediaConstraintsInterface::kValueFalse);
   constraints.AddMandatory(webrtc::MediaConstraintsInterface::kOfferToReceiveVideo, webrtc::MediaConstraintsInterface::kValueFalse);

-  _audioDeviceModule = new webrtc::FakeAudioDeviceModule();
-  _jinglePeerConnectionFactory = webrtc::CreatePeerConnectionFactory(_workerThread, _signalingThread, _audioDeviceModule, nullptr, nullptr);
+  if (!_jinglePeerConnectionFactory) {
+    auto audioDeviceModule = new webrtc::FakeAudioDeviceModule();
+    _jinglePeerConnectionFactory = webrtc::CreatePeerConnectionFactory(_workerThread, _signalingThread,
+                                                                       audioDeviceModule, nullptr, nullptr);
+  }
   _jinglePeerConnection = _jinglePeerConnectionFactory->CreatePeerConnection(configuration, &constraints, nullptr, nullptr, this);
   uv_mutex_init(&lock);
@@ -72,7 +76,6 @@ PeerConnection::PeerConnection(webrtc::PeerConnectionInterface::IceServers iceSe
 PeerConnection::~PeerConnection() {
   TRACE_CALL;
   _jinglePeerConnection = nullptr;
-  _jinglePeerConnectionFactory = nullptr;
   uv_mutex_destroy(&lock);
   TRACE_END;
 }
@@ -548,7 +551,6 @@ NAN_METHOD(PeerConnection::Close) {
   }
   self->_jinglePeerConnection = nullptr;
-  self->_jinglePeerConnectionFactory = nullptr;
   TRACE_END;
   info.GetReturnValue().Set(Nan::Undefined());
@@ -663,6 +665,10 @@ NAN_SETTER(PeerConnection::ReadOnly) {
   INFO("PeerConnection::ReadOnly");
 }
+void PeerConnection::Dispose() {
+  _jinglePeerConnectionFactory = nullptr;
+}
+
 void PeerConnection::Init(rtc::Thread* signalingThread, rtc::Thread* workerThread, Handle<Object> exports) {
   _signalingThread = signalingThread;
   _workerThread = workerThread;
diff --git a/src/peerconnection.h b/src/peerconnection.h
index 8edbe34..906d2e4 100644
--- a/src/peerconnection.h
+++ b/src/peerconnection.h
@@ -144,6 +144,7 @@ class PeerConnection
   // Nodejs wrapping.
   //
   static void Init(rtc::Thread* signalingThread, rtc::Thread* workerThread, v8::Handle<v8::Object> exports);
+  static void Dispose();
   static Nan::Persistent<v8::Function> constructor;
   static NAN_METHOD(New);
@@ -192,9 +193,8 @@ class PeerConnection
   rtc::scoped_refptr<SetLocalDescriptionObserver> _setLocalDescriptionObserver;
   rtc::scoped_refptr<SetRemoteDescriptionObserver> _setRemoteDescriptionObserver;
-  webrtc::AudioDeviceModule* _audioDeviceModule;
   rtc::scoped_refptr<webrtc::PeerConnectionInterface> _jinglePeerConnection;
-  rtc::scoped_refptr<webrtc::PeerConnectionFactoryInterface> _jinglePeerConnectionFactory;
+  static rtc::scoped_refptr<webrtc::PeerConnectionFactoryInterface> _jinglePeerConnectionFactory;
   static rtc::Thread* _signalingThread;
   static rtc::Thread* _workerThread;
```

</p>
</details>
<br>

* Average CPU Usage (%): 95.842
* Max CPU Usage (%): 149.9
* Time (s): 6.644

## develop with Dummy ADM and N PCFIs

<details><summary>Patch</summary>
<p>

```patch
diff --git a/src/peerconnection.cc b/src/peerconnection.cc
index b112f69..6dde7cc 100644
--- a/src/peerconnection.cc
+++ b/src/peerconnection.cc
@@ -10,7 +10,7 @@
 #include "webrtc/api/mediaconstraintsinterface.h"
 #include "webrtc/api/test/fakeconstraints.h"
 #include "webrtc/base/refcount.h"
-#include "webrtc/modules/audio_device/include/fake_audio_device.h"
+#include "webrtc/modules/audio_device/include/audio_device.h"
 
 #include "common.h"
 #include "create-answer-observer.h"
@@ -59,8 +59,8 @@ PeerConnection::PeerConnection(webrtc::PeerConnectionInterface::IceServers iceSe
   constraints.AddMandatory(webrtc::MediaConstraintsInterface::kOfferToReceiveAudio, webrtc::MediaConstraintsInterface::kValueFalse);
   constraints.AddMandatory(webrtc::MediaConstraintsInterface::kOfferToReceiveVideo, webrtc::MediaConstraintsInterface::kValueFalse);
 
-  _audioDeviceModule = new webrtc::FakeAudioDeviceModule();
-  _jinglePeerConnectionFactory = webrtc::CreatePeerConnectionFactory(_workerThread, _signalingThread, _audioDeviceModule, nullptr, nullptr);
+  auto audioDeviceModule = webrtc::AudioDeviceModule::Create(0, webrtc::AudioDeviceModule::AudioLayer::kDummyAudio);
+  _jinglePeerConnectionFactory = webrtc::CreatePeerConnectionFactory(_workerThread, _signalingThread, audioDeviceModule, nullptr, nullptr);
   _jinglePeerConnection = _jinglePeerConnectionFactory->CreatePeerConnection(configuration, &constraints, nullptr, nullptr, this);
 
   uv_mutex_init(&lock);
diff --git a/src/peerconnection.h b/src/peerconnection.h
index 8edbe34..d917dd7 100644
--- a/src/peerconnection.h
+++ b/src/peerconnection.h
@@ -192,7 +192,6 @@ class PeerConnection
   rtc::scoped_refptr<SetLocalDescriptionObserver> _setLocalDescriptionObserver;
   rtc::scoped_refptr<SetRemoteDescriptionObserver> _setRemoteDescriptionObserver;
 
-  webrtc::AudioDeviceModule* _audioDeviceModule;
   rtc::scoped_refptr<webrtc::PeerConnectionInterface> _jinglePeerConnection;
   rtc::scoped_refptr<webrtc::PeerConnectionFactoryInterface> _jinglePeerConnectionFactory;
 

```

</p>
</details>
<br>

I can't actually complete the test. node-webrtc takes more and more CPU, topping out at ~84 RTCPeerConnections.

## develop with Dummy ADM and 1 PCFI

<details><summary>Patch</summary>
<p>
  
```patch
diff --git a/src/binding.cc b/src/binding.cc
index 6109516..4000ab8 100644
--- a/src/binding.cc
+++ b/src/binding.cc
@@ -38,6 +38,8 @@ void setup(rtc::Thread* signalingThread, rtc::Thread* workerThread) {
 }
 
 void dispose(void* args) {
+  node_webrtc::PeerConnection::Dispose();
+
   workerThread.Stop();
 
   signalingThread.Stop();
diff --git a/src/peerconnection.cc b/src/peerconnection.cc
index b112f69..f779b52 100644
--- a/src/peerconnection.cc
+++ b/src/peerconnection.cc
@@ -10,7 +10,7 @@
 #include "webrtc/api/mediaconstraintsinterface.h"
 #include "webrtc/api/test/fakeconstraints.h"
 #include "webrtc/base/refcount.h"
-#include "webrtc/modules/audio_device/include/fake_audio_device.h"
+#include "webrtc/modules/audio_device/include/audio_device.h"
 
 #include "common.h"
 #include "create-answer-observer.h"
@@ -36,6 +36,7 @@ using v8::Value;
 using v8::Array;
 
 Nan::Persistent<Function> PeerConnection::constructor;
+rtc::scoped_refptr<webrtc::PeerConnectionFactoryInterface> PeerConnection::_jinglePeerConnectionFactory;
 rtc::Thread* PeerConnection::_signalingThread;
 rtc::Thread* PeerConnection::_workerThread;
 
@@ -59,8 +60,11 @@ PeerConnection::PeerConnection(webrtc::PeerConnectionInterface::IceServers iceSe
   constraints.AddMandatory(webrtc::MediaConstraintsInterface::kOfferToReceiveAudio, webrtc::MediaConstraintsInterface::kValueFalse);
   constraints.AddMandatory(webrtc::MediaConstraintsInterface::kOfferToReceiveVideo, webrtc::MediaConstraintsInterface::kValueFalse);
 
-  _audioDeviceModule = new webrtc::FakeAudioDeviceModule();
-  _jinglePeerConnectionFactory = webrtc::CreatePeerConnectionFactory(_workerThread, _signalingThread, _audioDeviceModule, nullptr, nullptr);
+  if (!_jinglePeerConnectionFactory) {
+    auto audioDeviceModule = webrtc::AudioDeviceModule::Create(0, webrtc::AudioDeviceModule::AudioLayer::kDummyAudio);
+    _jinglePeerConnectionFactory = webrtc::CreatePeerConnectionFactory(_workerThread, _signalingThread,
+                                                                       audioDeviceModule, nullptr, nullptr);
+  }
   _jinglePeerConnection = _jinglePeerConnectionFactory->CreatePeerConnection(configuration, &constraints, nullptr, nullptr, this);
 
   uv_mutex_init(&lock);
@@ -72,7 +76,6 @@ PeerConnection::PeerConnection(webrtc::PeerConnectionInterface::IceServers iceSe
 PeerConnection::~PeerConnection() {
   TRACE_CALL;
   _jinglePeerConnection = nullptr;
-  _jinglePeerConnectionFactory = nullptr;
   uv_mutex_destroy(&lock);
   TRACE_END;
 }
@@ -548,7 +551,6 @@ NAN_METHOD(PeerConnection::Close) {
   }
 
   self->_jinglePeerConnection = nullptr;
-  self->_jinglePeerConnectionFactory = nullptr;
 
   TRACE_END;
   info.GetReturnValue().Set(Nan::Undefined());
@@ -663,6 +665,10 @@ NAN_SETTER(PeerConnection::ReadOnly) {
   INFO("PeerConnection::ReadOnly");
 }
 
+void PeerConnection::Dispose() {
+  _jinglePeerConnectionFactory = nullptr;
+}
+
 void PeerConnection::Init(rtc::Thread* signalingThread, rtc::Thread* workerThread, Handle<Object> exports) {
   _signalingThread = signalingThread;
   _workerThread = workerThread;
diff --git a/src/peerconnection.h b/src/peerconnection.h
index 8edbe34..906d2e4 100644
--- a/src/peerconnection.h
+++ b/src/peerconnection.h
@@ -144,6 +144,7 @@ class PeerConnection
   // Nodejs wrapping.
   //
   static void Init(rtc::Thread* signalingThread, rtc::Thread* workerThread, v8::Handle<v8::Object> exports);
+  static void Dispose();
   static Nan::Persistent<v8::Function> constructor;
   static NAN_METHOD(New);
 
@@ -192,9 +193,8 @@ class PeerConnection
   rtc::scoped_refptr<SetLocalDescriptionObserver> _setLocalDescriptionObserver;
   rtc::scoped_refptr<SetRemoteDescriptionObserver> _setRemoteDescriptionObserver;
 
-  webrtc::AudioDeviceModule* _audioDeviceModule;
   rtc::scoped_refptr<webrtc::PeerConnectionInterface> _jinglePeerConnection;
-  rtc::scoped_refptr<webrtc::PeerConnectionFactoryInterface> _jinglePeerConnectionFactory;
+  static rtc::scoped_refptr<webrtc::PeerConnectionFactoryInterface> _jinglePeerConnectionFactory;
 
   static rtc::Thread* _signalingThread;
   static rtc::Thread* _workerThread;

```

</p>
</details>
<br>

* Average CPU Usage (%): 81.492
* Max CPU Usage (%): 125.24
* Time (s): 6.367

**Comment by markandrus** [comment]
Fixed in 0.0.66
